### PR TITLE
Build global ordinals terms bucket from matching ordinals

### LINKED ISSUE #30117: string terms is very slow when there are millions of buckets
elasticsearch 6.1.3

In GlobalOrdinalsStringTermsAggregator
When there are levels of aggregation， parent agg and valueCount both more than 100 thousands bucket

the loop may be explode
`for (long globalTermOrd = 0; globalTermOrd < valueCount; ++globalTermOrd) `

My temporary resolution is， loop by bucketOrds
```
            private static Field keysField;
            static {
                try {
                   keysField = LongHash.class.getDeclaredField("keys");
                   keysField.setAccessible(true);
                 } catch (Exception e) {
                LOGGER.error(e.getMessage(), e);
            }
     
           if (bucketOrds != null && bucketOrds.size() < valueCount && bucketCountThresholds.getMinDocCount() > 0 ) {
                try {
                    loop = false;
                    LongArray keys = ((LongArray) keysField.get(bucketOrds));
                    for (long i = 0; i < keys.size(); i++) {
                        //i: bucketOrd
                        int bucketDocCount = bucketDocCount(i);
                        if ( bucketDocCount == 0) {
                            continue;
                        }           
                        long globalTermOrd = keys.get(i);
                        if (includeExclude != null && !acceptedGlobalOrdinals.get(globalTermOrd)) {
                            continue;
                        }
                        otherDocCount += bucketDocCount;
                        spare.globalOrd = globalTermOrd;
                        spare.bucketOrd = i;
                        spare.docCount = bucketDocCount;
                        if (bucketCountThresholds.getShardMinDocCount() <= spare.docCount) {
                            spare = ordered.insertWithOverflow(spare);
                            if (spare == null) {
                                spare = new OrdBucket(-1, 0, null, showTermDocCountError, 0);
                            }
                        }

                    }

                } catch (IllegalAccessException e) {
                    LOGGER.error(e.getMessage(), e);
                    loop = true;
                }
            }
```

### PR DESCRIPTION
**Pull request by jimczi**
The global ordinals terms aggregator has an option to remap global ordinals to
dense ordinal that match the request. This mode is automatically picked when the terms
aggregator is a child of another bucket aggregator or when it needs to defer buckets to an
aggregation that is used in the ordering of the terms.
Though when building the final buckets, this aggregator loops over all possible global ordinals
rather than using the hash map that was built to remap the ordinals.
For fields with high cardinality this is highly inefficient and can lead to slow responses even
when the number of terms that match the query is low.
This change fixes this performance issue by using the hash table of matching ordinals to perform
the pruning of the final buckets for the terms and significant_terms aggregation.
I ran a simple benchmark with 1M documents containing 0 to 10 keywords randomly selected among 1M unique terms.
This field is used to perform a multi-level terms aggregation using rally to collect the response times.
The aggregation below is an example of a two-level terms aggregation that was used to perform the benchmark:

```
"aggregations":{
   "1":{
      "terms":{
         "field":"keyword"
      },
      "aggregations":{
         "2":{
            "terms":{
               "field":"keyword"
            }
         }
      }
   }
}
```

| Levels of aggregation | 50th percentile ms (master) | 50th percentile ms (patch) |
| --- | --- | --- |
| 2 | 640.41 | 577.499 |
| 3 | 2239.66 | 600.154 |
| 4 | 14141.2 | 703.512 |

Closes #30117

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/GlobalOrdinalsSignificantTermsAggregator.java [hunk 1]
```diff
@@ -103,11 +101,22 @@ public SignificantStringTerms buildAggregation(long owningBucketOrdinal) throws
 
         BucketSignificancePriorityQueue<SignificantStringTerms.Bucket> ordered = new BucketSignificancePriorityQueue<>(size);
         SignificantStringTerms.Bucket spare = null;
-        for (long globalTermOrd = 0; globalTermOrd < valueCount; ++globalTermOrd) {
-            if (includeExclude != null && !acceptedGlobalOrdinals.get(globalTermOrd)) {
+        boolean needsFullSan = bucketOrds == null || bucketCountThresholds.getMinDocCount() == 0;
```
Comment (markharwood) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/GlobalOrdinalsSignificantTermsAggregator.java]: Typo - "needsFullScan"

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/GlobalOrdinalsSignificantTermsAggregator.java [hunk 2]
```diff
@@ -103,11 +101,22 @@ public SignificantStringTerms buildAggregation(long owningBucketOrdinal) throws
 
         BucketSignificancePriorityQueue<SignificantStringTerms.Bucket> ordered = new BucketSignificancePriorityQueue<>(size);
         SignificantStringTerms.Bucket spare = null;
-        for (long globalTermOrd = 0; globalTermOrd < valueCount; ++globalTermOrd) {
-            if (includeExclude != null && !acceptedGlobalOrdinals.get(globalTermOrd)) {
+        boolean needsFullScan = bucketOrds == null || bucketCountThresholds.getMinDocCount() == 0;
+        long maxId = needsFullScan ? valueCount : bucketOrds.size();
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/GlobalOrdinalsSignificantTermsAggregator.java]: let's make both final

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java [hunk 3]
```diff
@@ -188,17 +184,28 @@ public InternalAggregation buildAggregation(long owningBucketOrdinal) throws IOE
         long otherDocCount = 0;
         BucketPriorityQueue<OrdBucket> ordered = new BucketPriorityQueue<>(size, order.comparator(this));
         OrdBucket spare = new OrdBucket(-1, 0, null, showTermDocCountError, 0);
-        for (long globalTermOrd = 0; globalTermOrd < valueCount; ++globalTermOrd) {
-            if (includeExclude != null && !acceptedGlobalOrdinals.get(globalTermOrd)) {
+        boolean needsFullScan = bucketOrds == null || bucketCountThresholds.getMinDocCount() == 0;
+        long maxId = needsFullScan ? valueCount : bucketOrds.size();
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java]: let's make them final?

**Review by jpountz — APPROVED** [review]
Thanks @jimczi !

**Comment by pakein** [comment]
Get globalOrd by `globalOrd = bucketOrds.get(bucketOrd)`
 if globalOrd == 0,  it may be 
 bucket exists and globalOrd=0 
 or
 bucket not exists
 when globalOrd =0 match , it seems the result my be duplicate ?

**Comment by jimczi** [comment]
We use `bucketOrds#get` on a bucket ord that is guaranteed to exist since we only iterate the existing slots in `LongHash` and the ids in the hash are dense. The returned long is undefined if the slot is empty so `0` does not indicate a non-existing bucket but we should be safe here since we always use the filled slot.

**Comment by pakein** [comment]
thanks @jimczi. I read the code carefully and it's my fault.

**Comment by mattweber** [comment]
Thank you @jimczi and @pakein!  

I came across this issue when investigating a performance issue for one of my customers aggregations and applied this patch to a forked terms aggregation in a plugin for ES `6.2.1`.  For our use-case which is 4-levels of terms aggregations against ~3M docs and the 4th level being a very high cardinality field we went from ~350s down to ~20s!!

Any chance of getting this in before `6.4.0` as it has the potential for such massive performance improvements with no side-affects?

**Comment by jimczi** [comment]
> Any chance of getting this in before 6.4.0 as it has the potential for such massive performance improvements with no side-affects?

It was too late for 6.3.0 and it's not a critical bug (though this is arguable ;) ) hence 6.4.0.

> For our use-case which is 4-levels of terms aggregations against ~3M docs and the 4th level being a very high cardinality field we went from ~350s down to ~20s!!

Nice numbers ! Though 20s for 3M docs seems quite high even with 4 levels. How many terms do you retrieve per level ?

**Comment by jimczi** [comment]
I discussed with @colings86 and we've decided to backport this pr in 6.3.1. It's really too late for 6.3.0 but if there is a 6.3.1 (which is not guaranteed) this pr will be included.

**Comment by mattweber** [comment]
@jimczi Thank you, `6.3.1` is perfectly fine and I understand it is not guaranteed to happen.

For our use-case, the first 3-levels are 100, and the 4th is unbounded (essentially all doc ids for the bucket).  Its this 4th level id collection that is the issue and I am working on trying to get them to remove this "requirement".

**Comment by fredgalvao** [comment]
Do {the tag swap + 6.3 commit} mean anything in regards to the possibility of having this in 6.3.0? Did anything change?
This is such a huge win and amazing optimization!

**Comment by jimczi** [comment]
@fredgalvao yes this change will be in 6.3.0

**Pull request merged**